.github/workflows/push.yml

```diff
@@ -1,47 +1,136 @@
+        uses: docker/setup-buildx-action@v1
+      -
+        name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ env.DOCKERHUB_USER }}
+          password: ${{ env.DOCKERHUB_TOKEN }}
+      -
+        name: Build and push
+        id: docker_build
+        uses: docker/build-push-action@v2
+        with:
+          push: true
+          tags: ${{ env.IMAGE }}:${{ env.IMAGE_TAG }}
+          cache-from: type=registry,ref=${{ env.IMAGE }}:buildcache
+          cache-to: type=registry,ref=${{ env.IMAGE }}:buildcache,mode=max
+
+  tests:
+    needs:
+      - set_version
+      - build_image
+    # only one parallel job allowed - used shared testing project
+    concurrency:
+      group: ci
+      cancel-in-progress: false
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out the repo
+        uses: actions/checkout@v2
+      - name: Load Docker image
+        uses: ./.github/actions/load-image
       - name: Tests
+        run: docker-compose run --rm -u "$UID:$GID" -e TARGET_VERSION dev make ci
+        env:
+          TARGET_VERSION: ${{ needs.set_version.outputs.version }}
+
+  cross-compile:
+    needs:
+      - set_version
+      - build_image
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out the repo
+        uses: actions/checkout@v2
+      - name: Load Docker image
+        uses: ./.github/actions/load-image
+      - name: Cross compile
         run: |
-          docker-compose run --rm -u "$UID:$GID" dev make ci
-      - name: Compile
-        run: |
-          docker-compose run --rm -u "$UID:$GID" dev make build
-      - name: Compile and Release
-        if: startsWith(github.ref, 'refs/tags/')
-        run: |
-          docker-compose run --rm -u "$UID:$GID" -e GITHUB_TOKEN -e TARGET_VERSION dev make release
+          mkdir -p /tmp/go-cache
+          docker-compose run --rm -u "$UID:$GID" -e TARGET_VERSION -v /tmp/go-cache:/tmp/go-cache dev make build
+        env:
+          TARGET_VERSION: ${{ needs.set_version.outputs.version }}
+      - name: Zip build cache
+        if: startsWith(github.ref, 'refs/tags/') && needs.set_version.outputs.is_semantic_tag == 'true'
+        run: zip -q -4 -r /tmp/go-cache.zip /tmp/go-cache
+      - name: Upload build cache
+        if: startsWith(github.ref, 'refs/tags/') && needs.set_version.outputs.is_semantic_tag == 'true'
+        uses: actions/upload-artifact@v2
+        with:
+          name: build-cache
+          path: /tmp/go-cache.zip
+          retention-days: 1
+
+  release:
+    needs:
+      - set_version
+      - build_image
+      - tests
+      - cross-compile
+    runs-on: ubuntu-latest
+    if: startsWith(github.ref, 'refs/tags/') && needs.set_version.outputs.is_semantic_tag == 'true'
+    steps:
+      - name: Check out the repo
+        uses: actions/checkout@v2
+      - name: Load Docker image
+        uses: ./.github/actions/load-image
+      - name: Load build cache
```

Comment:
Pred `release` sa cache stiahne.
A tak sa nemusi vsetko kompilovat znova 3 minuty.

Idealne by bolo, ak by sme do `goreleaser` mohli dostat priamo skompilovane binarky, ... a v osobitnom kroku spravit `release` to ale `goreleaser` nevie, .... tak si pomozeme aspon touto build cache.